{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Framework\n",
    "\n",
    "- Data - Palbociclib Data (Proteomics and GDSC2)\n",
    "- Preprocessing method - missing data imputation by lower quantile (0 to 0.25)\n",
    "- Feature selection (Filtering) - select features with pearson correlation >= |0.4|\n",
    "- Model - Elastic Net\n",
    "- Target variable used - AUC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Github\\ode-biomarker-project\\dynamic-marker\\lib\\site-packages\\tqdm\\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import os \n",
    "import DataFunctions as utils\n",
    "import Visualisation as vis\n",
    "from toolkit import * \n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "save_results_path = 'data//results//workbook-4-jun' \n",
    "\n",
    "if not os.path.exists(save_results_path):\n",
    "    os.makedirs(save_results_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Load data\n",
    "\n",
    "import pandas as pd\n",
    "import pickle\n",
    "\n",
    "# import GDSC2 drug response data using pickle\n",
    "\n",
    "with open('data/drug-response/GDSC2/cache_gdsc2.pkl', 'rb') as f:\n",
    "    gdsc2 = pickle.load(f)\n",
    "    gdsc2_info = pickle.load(f)\n",
    "\n",
    "# import CCLE gene expression data using pickle\n",
    "\n",
    "with open('data/gene-expression/CCLE_Public_22Q2/ccle_expression.pkl', 'rb') as f:\n",
    "    gene_entrez = pickle.load(f)\n",
    "    ccle = pickle.load(f)\n",
    "\n",
    "# import CCLE sample info data using pickle\n",
    "\n",
    "with open('data/gene-expression/CCLE_Public_22Q2/ccle_sample_info.pkl', 'rb') as f:\n",
    "    ccle_sample_info = pickle.load(f)\n",
    "\n",
    "# import STRING database using pickle\n",
    "\n",
    "with open('data/protein-interaction/STRING/string_df.pkl', 'rb') as f:\n",
    "    string_df = pickle.load(f)\n",
    "    string_df_info = pickle.load(f)\n",
    "    string_df_alias = pickle.load(f)\n",
    "\n",
    "\n",
    "# import proteomic expression\n",
    "with open('data/proteomic-expression/goncalves-2022-cell/goncalve_proteome_fillna_processed.pkl', 'rb') as f:\n",
    "    joined_full_protein_matrix = pickle.load(f)\n",
    "    joined_sin_peptile_exclusion_matrix = pickle.load(f)\n",
    "\n",
    "# import STRING database using pickle\n",
    "\n",
    "with open('data/protein-interaction/STRING/string_df.pkl', 'rb') as f:\n",
    "    string_df = pickle.load(f)\n",
    "    string_df_info = pickle.load(f)\n",
    "    string_df_alias = pickle.load(f)\n",
    "\n",
    "# open STRING to goncalves mapping file\n",
    "\n",
    "with open('data\\protein-interaction\\STRING\\goncalve_to_string_id_df.pkl', 'rb') as f:\n",
    "    goncalve_to_string_id_df = pickle.load(f)\n",
    "\n",
    "# open the cache for neighbourhood calculations\n",
    "\n",
    "with open('data/protein-interaction/STRING/palbociclib_nth_degree_neighbours.pkl', 'rb') as f:\n",
    "    nth_degree_neighbours = pickle.load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Config Parameters\n",
    "\n",
    "## Data \n",
    "\n",
    "# create the full dataset\n",
    "drug_selected = 'Palbociclib'\n",
    "\n",
    "## Preprocessing Method\n",
    "# Between fill by zero, impute by random variable between 0 to 1st quantile\n",
    "# Impute1stQuantile, Fill0, ImputeKNN\n",
    "preprocessing_method = 'Impute1stQuantile'\n",
    "\n",
    "# set threshold for feature removal (0.8 means remove features with > 80% missing values)\n",
    "feature_nan_removal_threshold = 0.8\n",
    "\n",
    "## Feature Selection Methods\n",
    "\n",
    "## Models \n",
    "## Target variable used\n",
    "\n",
    "target_variable = 'AUC'\n",
    "\n",
    "## Hyperparameters\n",
    "experiment_name = 'test'\n",
    "save_input = True\n",
    "save_output = True\n",
    "\n",
    "# hyperparameters\n",
    "max_gene_target_disance = 2 # specify the level of biological relevance to drug target(s)\n",
    "statistical_filter_size = 100 # can be optimized using global feature dropout testing\n",
    "monte_carlo_cross_validation_size = 50 # can be automatically optimized via rank impact assessment\n",
    "models_used = ['ElasticNet']\n",
    "models_hyperparameters = [{'alpha': 0.005}]\n",
    "\n",
    "# extra hyperparameters\n",
    "statistical_filter_threshold = 0.05 # currently not in use\n",
    "cv_split_size = 0.1\n",
    "\n",
    "# generated hyperparameters\n",
    "rng_seed_lists = []\n",
    "for j in range(monte_carlo_cross_validation_size):\n",
    "    rng_seed_lists.append(np.random.randint(100000))\n",
    "\n",
    "\n",
    "verbose = True\n",
    "max_feature_save_size = 1000\n",
    "data_collector = []\n",
    "conditions_to_test = ['network_f_regression_selection']\n",
    "conditions_to_get_feature_importance = [True,]\n",
    "matched_functions = [get_network_stat_features,]\n",
    "extra_args = [(nth_degree_neighbours, max_gene_target_disance, statistical_filter_size),]\n",
    "features_to_knockout = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_df = utils.create_joint_dataset_from_proteome_gdsc(drug_selected, joined_sin_peptile_exclusion_matrix, gdsc2, drug_value=target_variable)\n",
    "feature_data, label_data = utils.create_feature_and_label(data_df, label_name=target_variable)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## Test a simple model run \n",
    "\n",
    "# # split the data into train and test set\n",
    "\n",
    "# import Visualisation as vis\n",
    "# import matplotlib.pyplot as plt\n",
    "# from sklearn.metrics import mean_squared_error\n",
    "# from sklearn.model_selection import train_test_split\n",
    "\n",
    "# # import lasso regression model\n",
    "\n",
    "# from sklearn.linear_model import ElasticNet, ElasticNetCV, Lasso, LinearRegression\n",
    "\n",
    "# from sklearn.ensemble import RandomForestRegressor\n",
    "\n",
    "# # import feature selection\n",
    "\n",
    "# from sklearn.feature_selection import SelectKBest, f_regression\n",
    "\n",
    "# # import log transformation\n",
    "\n",
    "# from sklearn.preprocessing import FunctionTransformer\n",
    "\n",
    "# # import pipeline\n",
    "\n",
    "# from sklearn.pipeline import Pipeline\n",
    "\n",
    "# # import scaler\n",
    "\n",
    "# from sklearn.preprocessing import StandardScaler, MinMaxScaler\n",
    "\n",
    "# # squared transform\n",
    "\n",
    "# sq_transform = FunctionTransformer(lambda x: x**2, validate=True)\n",
    "\n",
    "# import numpy as np\n",
    "\n",
    "# X_train, X_test, y_train, y_test = train_test_split(feature_data, label_data, test_size=0.1,\n",
    "#                                                     random_state=np.random.randint(1000))\n",
    "\n",
    "# # # plot the label distribution\n",
    "\n",
    "# import seaborn as sns\n",
    "\n",
    "# # sns.histplot(y_train, kde=True)\n",
    "\n",
    "# # sns.histplot(y_train_t, kde=True)\n",
    "\n",
    "# # sns.histplot(y_train_mm, kde=True)\n",
    "# # use the pipeline to fit the model\n",
    "\n",
    "# model = RandomForestRegressor(n_jobs=4)\n",
    "\n",
    "# # create the pipeline\n",
    "\n",
    "# pipeline = Pipeline([('scaler', StandardScaler()), \n",
    "#                      ('feature_selection', SelectKBest(f_regression, k=200)),\n",
    "#                         ('model', RandomForestRegressor(n_jobs=4))])\n",
    "\n",
    "# pipeline2 = Pipeline([('feature_selection', SelectKBest(f_regression, k=200)),\n",
    "#                         ('model', RandomForestRegressor(n_jobs=4))])\n",
    "\n",
    "# pipeline3 = Pipeline([('imputer', FirstQuantileImputer()) ,\n",
    "#                       ('scaler', StandardScaler()), \n",
    "#                       ('feature_selection', SelectKBest(f_regression, k=200)),\n",
    "#                         ('model', RandomForestRegressor(n_jobs=4))])\n",
    "\n",
    "# pipeline4 = Pipeline([('imputer', FirstQuantileImputer()), \n",
    "#                       ('feature_selection', SelectKBest(f_regression, k=200)),\n",
    "#                         ('model', RandomForestRegressor(n_jobs=4))])\n",
    "\n",
    "# pipeline5 = Pipeline([('feature_selection', SelectKBest(f_regression, k=100)),\n",
    "#                         ('model', RandomForestRegressor(n_jobs=4))])\n",
    "\n",
    "# pipeline6 = Pipeline([('feature_selection', SelectKBest(f_regression, k=1)),\n",
    "#                         ('model', RandomForestRegressor(n_jobs=4))])\n",
    "\n",
    "# # fit the model\n",
    "\n",
    "# pipeline.fit(X_train, y_train)\n",
    "# pipeline2.fit(X_train, y_train)\n",
    "# pipeline3.fit(X_train, y_train)\n",
    "# pipeline4.fit(X_train, y_train)\n",
    "# pipeline5.fit(X_train, y_train)\n",
    "# pipeline6.fit(X_train, y_train)\n",
    "\n",
    "# # predict the test set\n",
    "\n",
    "# y_pred = pipeline.predict(X_test)\n",
    "# y_pred2 = pipeline2.predict(X_test)\n",
    "# y_pred3 = pipeline3.predict(X_test)\n",
    "# y_pred4 = pipeline4.predict(X_test)\n",
    "# y_pred5 = pipeline5.predict(X_test)\n",
    "# y_pred6 = pipeline6.predict(X_test)\n",
    "\n",
    "# # plot the predicted vs actual\n",
    "\n",
    "# fig, axes = plt.subplots(3, 2, figsize=(20, 24))\n",
    "\n",
    "# vis.plot_predictions_vs_actual_values(y_test, y_pred, \"With scaling\", 'Actual', 'Predicted', ax=axes[0][0])\n",
    "# vis.plot_predictions_vs_actual_values(y_test, y_pred2, \"Without scaling\", 'Actual', 'Predicted', ax=axes[0][1])\n",
    "# vis.plot_predictions_vs_actual_values(y_test, y_pred3, \"With imputation\", 'Actual', 'Predicted', ax=axes[1][0])\n",
    "# vis.plot_predictions_vs_actual_values(y_test, y_pred4, \"With imputation and without scaling\", 'Actual', 'Predicted', ax=axes[1][1])\n",
    "# vis.plot_predictions_vs_actual_values(y_test, y_pred5, \"100 features\", 'Actual', 'Predicted', ax=axes[2][0])\n",
    "# vis.plot_predictions_vs_actual_values(y_test, y_pred6, \"1 features\", 'Actual', 'Predicted', ax=axes[2][1])\n",
    "\n",
    "# plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "------------------ feature size: 1\n",
      "------------------ feature size: 11\n",
      "------------------ feature size: 21\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\Github\\ode-biomarker-project\\workbook-4-jun-2023.ipynb Cell 7\u001b[0m in \u001b[0;36m1\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m \u001b[39mfor\u001b[39;00m feature_size \u001b[39min\u001b[39;00m feature_size_tested:\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39m------------------ feature size: \u001b[39m\u001b[39m{\u001b[39;00mfeature_size\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m)\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=15'>16</a>\u001b[0m     feature_size_df \u001b[39m=\u001b[39m run_bulk_test([\u001b[39m'\u001b[39m\u001b[39mrandom\u001b[39m\u001b[39m'\u001b[39m], \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m                                     [\u001b[39mFalse\u001b[39;00m], \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m                                     [impute_with_random_selection], \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=18'>19</a>\u001b[0m                                     [(feature_size,)],\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m                                     models_used, models_hyperparameters,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m                                     rng_seed_lists, \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=21'>22</a>\u001b[0m                                     feature_data, label_data,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=22'>23</a>\u001b[0m                                     cv_split_size\u001b[39m=\u001b[39mcv_split_size, \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m                                     verbose\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m, \n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=24'>25</a>\u001b[0m                                     bulk_run_tag\u001b[39m=\u001b[39m\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mfeature_size_\u001b[39m\u001b[39m{\u001b[39;00mfeature_size\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=25'>26</a>\u001b[0m                                     save_output\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=26'>27</a>\u001b[0m                                     n_jobs\u001b[39m=\u001b[39m\u001b[39m4\u001b[39m)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m     \u001b[39m# create a column for feature size for each row\u001b[39;00m\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X11sZmlsZQ%3D%3D?line=28'>29</a>\u001b[0m     feature_size_df[\u001b[39m'\u001b[39m\u001b[39mfeature_size\u001b[39m\u001b[39m'\u001b[39m] \u001b[39m=\u001b[39m feature_size\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\toolkit.py:247\u001b[0m, in \u001b[0;36mrun_bulk_test\u001b[1;34m(conditions_to_test, conditions_to_get_feature_importance, matched_functions, extra_args, models_used, models_hyperparameters, rng_seed_lists, feature_data, label_data, cv_split_size, max_feature_save_size, n_jobs, verbose, save_output, bulk_run_tag, output_file_path)\u001b[0m\n\u001b[0;32m    244\u001b[0m                 data_collector\u001b[39m.\u001b[39mappend(data)\n\u001b[0;32m    245\u001b[0m \u001b[39melse\u001b[39;00m: \n\u001b[0;32m    246\u001b[0m     \u001b[39m# use joblib to parallelize the process\u001b[39;00m\n\u001b[1;32m--> 247\u001b[0m     data_collector \u001b[39m=\u001b[39m Parallel(n_jobs\u001b[39m=\u001b[39;49mn_jobs)(delayed(run_single_test)(condition,\n\u001b[0;32m    248\u001b[0m                                     conditions_to_get_feature_importance[j],\n\u001b[0;32m    249\u001b[0m                                     matched_functions[j],\n\u001b[0;32m    250\u001b[0m                                     extra_args[j],\n\u001b[0;32m    251\u001b[0m                                     model_str,\n\u001b[0;32m    252\u001b[0m                                     models_hyperparameters[j],\n\u001b[0;32m    253\u001b[0m                                     rng,\n\u001b[0;32m    254\u001b[0m                                     feature_data, label_data,\n\u001b[0;32m    255\u001b[0m                                     cv_split_size,\n\u001b[0;32m    256\u001b[0m                                     max_feature_save_size,\n\u001b[0;32m    257\u001b[0m                                     verbose\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m) \n\u001b[0;32m    258\u001b[0m                                     \u001b[39mfor\u001b[39;49;00m model_str \u001b[39min\u001b[39;49;00m models_used\n\u001b[0;32m    259\u001b[0m                                     \u001b[39mfor\u001b[39;49;00m rng \u001b[39min\u001b[39;49;00m rng_seed_lists\n\u001b[0;32m    260\u001b[0m                                     \u001b[39mfor\u001b[39;49;00m j, condition \u001b[39min\u001b[39;49;00m \u001b[39menumerate\u001b[39;49m(conditions_to_test))\n\u001b[0;32m    263\u001b[0m \u001b[39mif\u001b[39;00m verbose:\n\u001b[0;32m    264\u001b[0m     \u001b[39mprint\u001b[39m(\u001b[39m'\u001b[39m\u001b[39m### All models ran\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\dynamic-marker\\lib\\site-packages\\joblib\\parallel.py:1098\u001b[0m, in \u001b[0;36mParallel.__call__\u001b[1;34m(self, iterable)\u001b[0m\n\u001b[0;32m   1095\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_iterating \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m\n\u001b[0;32m   1097\u001b[0m \u001b[39mwith\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backend\u001b[39m.\u001b[39mretrieval_context():\n\u001b[1;32m-> 1098\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mretrieve()\n\u001b[0;32m   1099\u001b[0m \u001b[39m# Make sure that we get a last message telling us we are done\u001b[39;00m\n\u001b[0;32m   1100\u001b[0m elapsed_time \u001b[39m=\u001b[39m time\u001b[39m.\u001b[39mtime() \u001b[39m-\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_start_time\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\dynamic-marker\\lib\\site-packages\\joblib\\parallel.py:975\u001b[0m, in \u001b[0;36mParallel.retrieve\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    973\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m    974\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mgetattr\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_backend, \u001b[39m'\u001b[39m\u001b[39msupports_timeout\u001b[39m\u001b[39m'\u001b[39m, \u001b[39mFalse\u001b[39;00m):\n\u001b[1;32m--> 975\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_output\u001b[39m.\u001b[39mextend(job\u001b[39m.\u001b[39;49mget(timeout\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mtimeout))\n\u001b[0;32m    976\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    977\u001b[0m         \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_output\u001b[39m.\u001b[39mextend(job\u001b[39m.\u001b[39mget())\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\dynamic-marker\\lib\\site-packages\\joblib\\_parallel_backends.py:567\u001b[0m, in \u001b[0;36mLokyBackend.wrap_future_result\u001b[1;34m(future, timeout)\u001b[0m\n\u001b[0;32m    564\u001b[0m \u001b[39m\"\"\"Wrapper for Future.result to implement the same behaviour as\u001b[39;00m\n\u001b[0;32m    565\u001b[0m \u001b[39mAsyncResults.get from multiprocessing.\"\"\"\u001b[39;00m\n\u001b[0;32m    566\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m--> 567\u001b[0m     \u001b[39mreturn\u001b[39;00m future\u001b[39m.\u001b[39;49mresult(timeout\u001b[39m=\u001b[39;49mtimeout)\n\u001b[0;32m    568\u001b[0m \u001b[39mexcept\u001b[39;00m CfTimeoutError \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m    569\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTimeoutError\u001b[39;00m \u001b[39mfrom\u001b[39;00m \u001b[39me\u001b[39;00m\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python38\\lib\\concurrent\\futures\\_base.py:434\u001b[0m, in \u001b[0;36mFuture.result\u001b[1;34m(self, timeout)\u001b[0m\n\u001b[0;32m    431\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_state \u001b[39m==\u001b[39m FINISHED:\n\u001b[0;32m    432\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m__get_result()\n\u001b[1;32m--> 434\u001b[0m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_condition\u001b[39m.\u001b[39;49mwait(timeout)\n\u001b[0;32m    436\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_state \u001b[39min\u001b[39;00m [CANCELLED, CANCELLED_AND_NOTIFIED]:\n\u001b[0;32m    437\u001b[0m     \u001b[39mraise\u001b[39;00m CancelledError()\n",
      "File \u001b[1;32m~\\AppData\\Local\\Programs\\Python\\Python38\\lib\\threading.py:302\u001b[0m, in \u001b[0;36mCondition.wait\u001b[1;34m(self, timeout)\u001b[0m\n\u001b[0;32m    300\u001b[0m \u001b[39mtry\u001b[39;00m:    \u001b[39m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[39;00m\n\u001b[0;32m    301\u001b[0m     \u001b[39mif\u001b[39;00m timeout \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m--> 302\u001b[0m         waiter\u001b[39m.\u001b[39;49macquire()\n\u001b[0;32m    303\u001b[0m         gotit \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n\u001b[0;32m    304\u001b[0m     \u001b[39melse\u001b[39;00m:\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "## Test Feature Size Effect on training performance between 0 to 500 \n",
    "\n",
    "def impute_with_random_selection(X_train, y_train, X_test, n_features):\n",
    "    X_train, y_train, X_test = impute_by_first_quantile(X_train, y_train, X_test)\n",
    "    X_train, y_train, X_test = get_random_features(X_train, y_train, X_test, n_features)\n",
    "    return X_train, y_train, X_test\n",
    "\n",
    "feature_size_tested = []\n",
    "for i in range(1, 500, 10):\n",
    "    feature_size_tested.append(i)\n",
    "\n",
    "all_df = pd.DataFrame()\n",
    "\n",
    "for feature_size in feature_size_tested:\n",
    "    print(f'------------------ feature size: {feature_size}')\n",
    "    feature_size_df = run_bulk_test(['random'], \n",
    "                                    [False], \n",
    "                                    [impute_with_random_selection], \n",
    "                                    [(feature_size,)],\n",
    "                                    models_used, models_hyperparameters,\n",
    "                                    rng_seed_lists, \n",
    "                                    feature_data, label_data,\n",
    "                                    cv_split_size=cv_split_size, \n",
    "                                    verbose=False, \n",
    "                                    bulk_run_tag=f'feature_size_{feature_size}',\n",
    "                                    save_output=False,\n",
    "                                    n_jobs=4)\n",
    "    # create a column for feature size for each row\n",
    "    feature_size_df['feature_size'] = feature_size\n",
    "    all_df = pd.concat([all_df, feature_size_df], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2500, 14)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'y'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py:3800\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3799\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m-> 3800\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_engine\u001b[39m.\u001b[39;49mget_loc(casted_key)\n\u001b[0;32m   3801\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\pandas\\_libs\\index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\pandas\\_libs\\index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\_libs\\hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'y'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[1;32mc:\\Github\\ode-biomarker-project\\workbook-4-jun-2023.ipynb Cell 8\u001b[0m in \u001b[0;36m8\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X10sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m \u001b[39m# plot a line plot with error bars for pearson correlation\u001b[39;00m\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X10sZmlsZQ%3D%3D?line=6'>7</a>\u001b[0m fig, ax \u001b[39m=\u001b[39m plt\u001b[39m.\u001b[39msubplots(figsize\u001b[39m=\u001b[39m(\u001b[39m14\u001b[39m, \u001b[39m7\u001b[39m))\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X10sZmlsZQ%3D%3D?line=7'>8</a>\u001b[0m sns\u001b[39m.\u001b[39mlineplot(x\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mfeature_size\u001b[39m\u001b[39m'\u001b[39m, y\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mcorr\u001b[39m\u001b[39m'\u001b[39m, data\u001b[39m=\u001b[39mall_df, ax\u001b[39m=\u001b[39max)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X10sZmlsZQ%3D%3D?line=8'>9</a>\u001b[0m ax\u001b[39m.\u001b[39mset_title(\u001b[39mf\u001b[39m\u001b[39m'\u001b[39m\u001b[39mPearson correlation vs feature size\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Github/ode-biomarker-project/workbook-4-jun-2023.ipynb#X10sZmlsZQ%3D%3D?line=9'>10</a>\u001b[0m ax\u001b[39m.\u001b[39mset_xlabel(\u001b[39m'\u001b[39m\u001b[39mfeature size\u001b[39m\u001b[39m'\u001b[39m)\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\seaborn\\relational.py:645\u001b[0m, in \u001b[0;36mlineplot\u001b[1;34m(data, x, y, hue, size, style, units, palette, hue_order, hue_norm, sizes, size_order, size_norm, dashes, markers, style_order, estimator, errorbar, n_boot, seed, orient, sort, err_style, err_kws, legend, ci, ax, **kwargs)\u001b[0m\n\u001b[0;32m    642\u001b[0m color \u001b[39m=\u001b[39m kwargs\u001b[39m.\u001b[39mpop(\u001b[39m\"\u001b[39m\u001b[39mcolor\u001b[39m\u001b[39m\"\u001b[39m, kwargs\u001b[39m.\u001b[39mpop(\u001b[39m\"\u001b[39m\u001b[39mc\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m))\n\u001b[0;32m    643\u001b[0m kwargs[\u001b[39m\"\u001b[39m\u001b[39mcolor\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m _default_color(ax\u001b[39m.\u001b[39mplot, hue, color, kwargs)\n\u001b[1;32m--> 645\u001b[0m p\u001b[39m.\u001b[39;49mplot(ax, kwargs)\n\u001b[0;32m    646\u001b[0m \u001b[39mreturn\u001b[39;00m ax\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\seaborn\\relational.py:441\u001b[0m, in \u001b[0;36m_LinePlotter.plot\u001b[1;34m(self, ax, kws)\u001b[0m\n\u001b[0;32m    438\u001b[0m     grouped \u001b[39m=\u001b[39m sub_data\u001b[39m.\u001b[39mgroupby(orient, sort\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39msort)\n\u001b[0;32m    439\u001b[0m     \u001b[39m# Could pass as_index=False instead of reset_index,\u001b[39;00m\n\u001b[0;32m    440\u001b[0m     \u001b[39m# but that fails on a corner case with older pandas.\u001b[39;00m\n\u001b[1;32m--> 441\u001b[0m     sub_data \u001b[39m=\u001b[39m grouped\u001b[39m.\u001b[39;49mapply(agg, other)\u001b[39m.\u001b[39mreset_index()\n\u001b[0;32m    442\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m    443\u001b[0m     sub_data[\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mother\u001b[39m}\u001b[39;00m\u001b[39mmin\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m np\u001b[39m.\u001b[39mnan\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\pandas\\core\\groupby\\groupby.py:1558\u001b[0m, in \u001b[0;36mGroupBy.apply\u001b[1;34m(self, func, *args, **kwargs)\u001b[0m\n\u001b[0;32m   1556\u001b[0m \u001b[39mwith\u001b[39;00m option_context(\u001b[39m\"\u001b[39m\u001b[39mmode.chained_assignment\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mNone\u001b[39;00m):\n\u001b[0;32m   1557\u001b[0m     \u001b[39mtry\u001b[39;00m:\n\u001b[1;32m-> 1558\u001b[0m         result \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_python_apply_general(f, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_selected_obj)\n\u001b[0;32m   1559\u001b[0m     \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[0;32m   1560\u001b[0m         \u001b[39m# gh-20949\u001b[39;00m\n\u001b[0;32m   1561\u001b[0m         \u001b[39m# try again, with .apply acting as a filtering\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1565\u001b[0m         \u001b[39m# fails on *some* columns, e.g. a numeric operation\u001b[39;00m\n\u001b[0;32m   1566\u001b[0m         \u001b[39m# on a string grouper column\u001b[39;00m\n\u001b[0;32m   1568\u001b[0m         \u001b[39mwith\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_group_selection_context():\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\pandas\\core\\groupby\\groupby.py:1610\u001b[0m, in \u001b[0;36mGroupBy._python_apply_general\u001b[1;34m(self, f, data, not_indexed_same, is_transform, is_agg)\u001b[0m\n\u001b[0;32m   1573\u001b[0m \u001b[39m@final\u001b[39m\n\u001b[0;32m   1574\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_python_apply_general\u001b[39m(\n\u001b[0;32m   1575\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1580\u001b[0m     is_agg: \u001b[39mbool\u001b[39m \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m,\n\u001b[0;32m   1581\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m NDFrameT:\n\u001b[0;32m   1582\u001b[0m     \u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m   1583\u001b[0m \u001b[39m    Apply function f in python space\u001b[39;00m\n\u001b[0;32m   1584\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   1608\u001b[0m \u001b[39m        data after applying f\u001b[39;00m\n\u001b[0;32m   1609\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[1;32m-> 1610\u001b[0m     values, mutated \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mgrouper\u001b[39m.\u001b[39;49mapply(f, data, \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49maxis)\n\u001b[0;32m   1611\u001b[0m     \u001b[39mif\u001b[39;00m not_indexed_same \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m   1612\u001b[0m         not_indexed_same \u001b[39m=\u001b[39m mutated \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mmutated\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\pandas\\core\\groupby\\ops.py:839\u001b[0m, in \u001b[0;36mBaseGrouper.apply\u001b[1;34m(self, f, data, axis)\u001b[0m\n\u001b[0;32m    837\u001b[0m \u001b[39m# group might be modified\u001b[39;00m\n\u001b[0;32m    838\u001b[0m group_axes \u001b[39m=\u001b[39m group\u001b[39m.\u001b[39maxes\n\u001b[1;32m--> 839\u001b[0m res \u001b[39m=\u001b[39m f(group)\n\u001b[0;32m    840\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m mutated \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m _is_indexed_like(res, group_axes, axis):\n\u001b[0;32m    841\u001b[0m     mutated \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\pandas\\core\\groupby\\groupby.py:1541\u001b[0m, in \u001b[0;36mGroupBy.apply.<locals>.f\u001b[1;34m(g)\u001b[0m\n\u001b[0;32m   1538\u001b[0m \u001b[39m@wraps\u001b[39m(func)\n\u001b[0;32m   1539\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mf\u001b[39m(g):\n\u001b[0;32m   1540\u001b[0m     \u001b[39mwith\u001b[39;00m np\u001b[39m.\u001b[39merrstate(\u001b[39mall\u001b[39m\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mignore\u001b[39m\u001b[39m\"\u001b[39m):\n\u001b[1;32m-> 1541\u001b[0m         \u001b[39mreturn\u001b[39;00m func(g, \u001b[39m*\u001b[39;49margs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\seaborn\\_statistics.py:482\u001b[0m, in \u001b[0;36mEstimateAggregator.__call__\u001b[1;34m(self, data, var)\u001b[0m\n\u001b[0;32m    480\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__call__\u001b[39m(\u001b[39mself\u001b[39m, data, var):\n\u001b[0;32m    481\u001b[0m     \u001b[39m\"\"\"Aggregate over `var` column of `data` with estimate and error interval.\"\"\"\u001b[39;00m\n\u001b[1;32m--> 482\u001b[0m     vals \u001b[39m=\u001b[39m data[var]\n\u001b[0;32m    483\u001b[0m     \u001b[39mif\u001b[39;00m callable(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mestimator):\n\u001b[0;32m    484\u001b[0m         \u001b[39m# You would think we could pass to vals.agg, and yet:\u001b[39;00m\n\u001b[0;32m    485\u001b[0m         \u001b[39m# https://github.com/mwaskom/seaborn/issues/2943\u001b[39;00m\n\u001b[0;32m    486\u001b[0m         estimate \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mestimator(vals)\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\pandas\\core\\frame.py:3805\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3803\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcolumns\u001b[39m.\u001b[39mnlevels \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[0;32m   3804\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_getitem_multilevel(key)\n\u001b[1;32m-> 3805\u001b[0m indexer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcolumns\u001b[39m.\u001b[39;49mget_loc(key)\n\u001b[0;32m   3806\u001b[0m \u001b[39mif\u001b[39;00m is_integer(indexer):\n\u001b[0;32m   3807\u001b[0m     indexer \u001b[39m=\u001b[39m [indexer]\n",
      "File \u001b[1;32mc:\\Github\\ode-biomarker-project\\venv\\lib\\site-packages\\pandas\\core\\indexes\\base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key, method, tolerance)\u001b[0m\n\u001b[0;32m   3800\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_engine\u001b[39m.\u001b[39mget_loc(casted_key)\n\u001b[0;32m   3801\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mKeyError\u001b[39;00m \u001b[39mas\u001b[39;00m err:\n\u001b[1;32m-> 3802\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mKeyError\u001b[39;00m(key) \u001b[39mfrom\u001b[39;00m \u001b[39merr\u001b[39;00m\n\u001b[0;32m   3803\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mTypeError\u001b[39;00m:\n\u001b[0;32m   3804\u001b[0m     \u001b[39m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[0;32m   3805\u001b[0m     \u001b[39m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[0;32m   3806\u001b[0m     \u001b[39m#  the TypeError.\u001b[39;00m\n\u001b[0;32m   3807\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'y'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABHsAAAJMCAYAAABq/qrBAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjYuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8o6BhiAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAl3ElEQVR4nO3df2zW5b34/1dboNXMVjyM8mPdYcf9cAsKDrSrznNi0tlkhnP4YzkMFyAcnXEHjdKzcwBFOudGOZsaloAjMhfPPxw4M5MsQupxPSM7HptD5EeiOYBxjJUYW+AstJy6Udfenz/2Xf12lB93Rwt9+Xgk9x9cu677fd1LLjFP3/f9LikUCoUAAAAAIIXSS70BAAAAAC4esQcAAAAgEbEHAAAAIBGxBwAAACARsQcAAAAgEbEHAAAAIBGxBwAAACARsQcAAAAgEbEHAAAAIBGxBwAAACCRomPPz3/+85g3b15MmzYtSkpKYvv27edds2vXrvjsZz8b5eXl8fGPfzyee+65YWwVAAAAgPMpOvb09PTErFmzYuPGjRc0/5e//GXceeedcfvtt8f+/fvjoYceinvuuSdeeumlojcLAAAAwLmVFAqFwrAXl5TECy+8EPPnzz/rnBUrVsSOHTvijTfeGBj78pe/HCdPnoyWlpbhXhoAAACAIYwb6Qu0tbVFfX39oLGGhoZ46KGHzrrm9OnTcfr06YE/9/f3x69//ev4sz/7sygpKRmprQIAAACMqkKhEKdOnYpp06ZFaenF+WnlEY89HR0dUV1dPWisuro6uru74ze/+U1cccUVZ6xpbm6Oxx57bKS3BgAAAHBZOHr0aHzkIx+5KO814rFnOFatWhWNjY0Df+7q6oqPfvSjcfTo0aisrLyEOwMAAAC4eLq7u6Ompiauuuqqi/aeIx57pkyZEp2dnYPGOjs7o7Kycsi7eiIiysvLo7y8/IzxyspKsQcAAABI52L+bM3F+TLYOdTV1UVra+ugsZdffjnq6upG+tIAAAAAHzhFx57/+7//i/3798f+/fsj4vePVt+/f3+0t7dHxO+/grV48eKB+ffdd18cPnw4/umf/ikOHjwYTz/9dPzbv/1bLF++/OJ8AgAAAAAGFB17XnvttbjxxhvjxhtvjIiIxsbGuPHGG2PNmjUREfHOO+8MhJ+IiI997GOxY8eOePnll2PWrFnx5JNPxg9+8INoaGi4SB8BAAAAgD8oKRQKhUu9ifPp7u6Oqqqq6Orq8ps9AAAAQBoj0TxG/Dd7AAAAABg9Yg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiYg8AAABAImIPAAAAQCJiDwAAAEAiw4o9GzdujBkzZkRFRUXU1tbG7t27zzl//fr18alPfSquuOKKqKmpieXLl8dvf/vbYW0YAAAAgLMrOvZs27YtGhsbo6mpKfbu3RuzZs2KhoaGOHbs2JDzt2zZEitXroympqY4cOBAPPvss7Ft27Z4+OGH/+TNAwAAADBY0bHnqaeeiq9+9auxdOnS+MxnPhObNm2KK6+8Mn74wx8OOf/VV1+NW2+9Ne66666YMWNG3HHHHbFw4cLz3g0EAAAAQPGKij29vb2xZ8+eqK+vf/8NSkujvr4+2trahlxzyy23xJ49ewbizuHDh2Pnzp3xxS9+8azXOX36dHR3dw96AQAAAHB+44qZfOLEiejr64vq6upB49XV1XHw4MEh19x1111x4sSJ+PznPx+FQiF+97vfxX333XfOr3E1NzfHY489VszWAAAAAIhReBrXrl27Yu3atfH000/H3r1748c//nHs2LEjHn/88bOuWbVqVXR1dQ28jh49OtLbBAAAAEihqDt7Jk2aFGVlZdHZ2TlovLOzM6ZMmTLkmkcffTQWLVoU99xzT0REXH/99dHT0xP33ntvPPLII1FaemZvKi8vj/Ly8mK2BgAAAEAUeWfPhAkTYs6cOdHa2jow1t/fH62trVFXVzfkmnffffeMoFNWVhYREYVCodj9AgAAAHAORd3ZExHR2NgYS5Ysiblz58bNN98c69evj56enli6dGlERCxevDimT58ezc3NERExb968eOqpp+LGG2+M2traeOutt+LRRx+NefPmDUQfAAAAAC6OomPPggUL4vjx47FmzZro6OiI2bNnR0tLy8CPNre3tw+6k2f16tVRUlISq1evjrfffjs+/OEPx7x58+Lb3/72xfsUAAAAAERERElhDHyXqru7O6qqqqKrqysqKysv9XYAAAAALoqRaB4j/jQuAAAAAEaP2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJCI2AMAAACQiNgDAAAAkIjYAwAAAJDIsGLPxo0bY8aMGVFRURG1tbWxe/fuc84/efJkLFu2LKZOnRrl5eXxyU9+Mnbu3DmsDQMAAABwduOKXbBt27ZobGyMTZs2RW1tbaxfvz4aGhri0KFDMXny5DPm9/b2xhe+8IWYPHlyPP/88zF9+vT41a9+FVdfffXF2D8AAAAA/z8lhUKhUMyC2trauOmmm2LDhg0REdHf3x81NTXxwAMPxMqVK8+Yv2nTpvjud78bBw8ejPHjxw9rk93d3VFVVRVdXV1RWVk5rPcAAAAAuNyMRPMo6mtcvb29sWfPnqivr3//DUpLo76+Ptra2oZc85Of/CTq6upi2bJlUV1dHTNnzoy1a9dGX1/fWa9z+vTp6O7uHvQCAAAA4PyKij0nTpyIvr6+qK6uHjReXV0dHR0dQ645fPhwPP/889HX1xc7d+6MRx99NJ588sn41re+ddbrNDc3R1VV1cCrpqammG0CAAAAfGCN+NO4+vv7Y/LkyfHMM8/EnDlzYsGCBfHII4/Epk2bzrpm1apV0dXVNfA6evToSG8TAAAAIIWifqB50qRJUVZWFp2dnYPGOzs7Y8qUKUOumTp1aowfPz7KysoGxj796U9HR0dH9Pb2xoQJE85YU15eHuXl5cVsDQAAAIAo8s6eCRMmxJw5c6K1tXVgrL+/P1pbW6Ourm7INbfeemu89dZb0d/fPzD25ptvxtSpU4cMPQAAAAAMX9Ff42psbIzNmzfHv/zLv8SBAwfia1/7WvT09MTSpUsjImLx4sWxatWqgflf+9rX4te//nU8+OCD8eabb8aOHTti7dq1sWzZsov3KQAAAACIiCK/xhURsWDBgjh+/HisWbMmOjo6Yvbs2dHS0jLwo83t7e1RWvp+Q6qpqYmXXnopli9fHjfccENMnz49HnzwwVixYsXF+xQAAAAARERESaFQKFzqTZzPSDxzHgAAAOBSG4nmMeJP4wIAAABg9Ig9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJiD0AAAAAiYg9AAAAAImIPQAAAACJDCv2bNy4MWbMmBEVFRVRW1sbu3fvvqB1W7dujZKSkpg/f/5wLgsAAADAeRQde7Zt2xaNjY3R1NQUe/fujVmzZkVDQ0McO3bsnOuOHDkSX//61+O2224b9mYBAAAAOLeiY89TTz0VX/3qV2Pp0qXxmc98JjZt2hRXXnll/PCHPzzrmr6+vvjKV74Sjz32WPzFX/zFn7RhAAAAAM6uqNjT29sbe/bsifr6+vffoLQ06uvro62t7azrvvnNb8bkyZPj7rvvHv5OAQAAADivccVMPnHiRPT19UV1dfWg8erq6jh48OCQa1555ZV49tlnY//+/Rd8ndOnT8fp06cH/tzd3V3MNgEAAAA+sEb0aVynTp2KRYsWxebNm2PSpEkXvK65uTmqqqoGXjU1NSO4SwAAAIA8irqzZ9KkSVFWVhadnZ2Dxjs7O2PKlClnzP/FL34RR44ciXnz5g2M9ff3//7C48bFoUOH4tprrz1j3apVq6KxsXHgz93d3YIPAAAAwAUoKvZMmDAh5syZE62trQOPT+/v74/W1ta4//77z5h/3XXXxeuvvz5obPXq1XHq1Kn43ve+d9aAU15eHuXl5cVsDQAAAIAoMvZERDQ2NsaSJUti7ty5cfPNN8f69eujp6cnli5dGhERixcvjunTp0dzc3NUVFTEzJkzB62/+uqrIyLOGAcAAADgT1d07FmwYEEcP3481qxZEx0dHTF79uxoaWkZ+NHm9vb2KC0d0Z8CAgAAAOAsSgqFQuFSb+J8uru7o6qqKrq6uqKysvJSbwcAAADgohiJ5uEWHAAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBEhhV7Nm7cGDNmzIiKioqora2N3bt3n3Xu5s2b47bbbouJEyfGxIkTo76+/pzzAQAAABi+omPPtm3borGxMZqammLv3r0xa9asaGhoiGPHjg05f9euXbFw4cL42c9+Fm1tbVFTUxN33HFHvP3223/y5gEAAAAYrKRQKBSKWVBbWxs33XRTbNiwISIi+vv7o6amJh544IFYuXLledf39fXFxIkTY8OGDbF48eILumZ3d3dUVVVFV1dXVFZWFrNdAAAAgMvWSDSPou7s6e3tjT179kR9ff37b1BaGvX19dHW1nZB7/Huu+/Ge++9F9dcc81Z55w+fTq6u7sHvQAAAAA4v6Jiz4kTJ6Kvry+qq6sHjVdXV0dHR8cFvceKFSti2rRpg4LRH2tubo6qqqqBV01NTTHbBAAAAPjAGtWnca1bty62bt0aL7zwQlRUVJx13qpVq6Krq2vgdfTo0VHcJQAAAMDYNa6YyZMmTYqysrLo7OwcNN7Z2RlTpkw559onnngi1q1bFz/96U/jhhtuOOfc8vLyKC8vL2ZrAAAAAESRd/ZMmDAh5syZE62trQNj/f390draGnV1dWdd953vfCcef/zxaGlpiblz5w5/twAAAACcU1F39kRENDY2xpIlS2Lu3Llx8803x/r166OnpyeWLl0aERGLFy+O6dOnR3Nzc0RE/PM//3OsWbMmtmzZEjNmzBj4bZ8PfehD8aEPfegifhQAAAAAio49CxYsiOPHj8eaNWuio6MjZs+eHS0tLQM/2tze3h6lpe/fMPT9738/ent740tf+tKg92lqaopvfOMbf9ruAQAAABikpFAoFC71Js5nJJ45DwAAAHCpjUTzGNWncQEAAAAwssQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARMQeAAAAgETEHgAAAIBExB4AAACARIYVezZu3BgzZsyIioqKqK2tjd27d59z/o9+9KO47rrroqKiIq6//vrYuXPnsDYLAAAAwLkVHXu2bdsWjY2N0dTUFHv37o1Zs2ZFQ0NDHDt2bMj5r776aixcuDDuvvvu2LdvX8yfPz/mz58fb7zxxp+8eQAAAAAGKykUCoViFtTW1sZNN90UGzZsiIiI/v7+qKmpiQceeCBWrlx5xvwFCxZET09PvPjiiwNjn/vc52L27NmxadOmC7pmd3d3VFVVRVdXV1RWVhazXQAAAIDL1kg0j3HFTO7t7Y09e/bEqlWrBsZKS0ujvr4+2trahlzT1tYWjY2Ng8YaGhpi+/btZ73O6dOn4/Tp0wN/7urqiojf/x8AAAAAkMUfWkeR9+KcU1Gx58SJE9HX1xfV1dWDxqurq+PgwYNDruno6BhyfkdHx1mv09zcHI899tgZ4zU1NcVsFwAAAGBM+N///d+oqqq6KO9VVOwZLatWrRp0N9DJkyfjz//8z6O9vf2ifXDgfd3d3VFTUxNHjx71VUkYAc4YjCxnDEaWMwYjq6urKz760Y/GNddcc9Hes6jYM2nSpCgrK4vOzs5B452dnTFlypQh10yZMqWo+RER5eXlUV5efsZ4VVWVf7jACKqsrHTGYAQ5YzCynDEYWc4YjKzS0mE9MH3o9ypm8oQJE2LOnDnR2to6MNbf3x+tra1RV1c35Jq6urpB8yMiXn755bPOBwAAAGD4iv4aV2NjYyxZsiTmzp0bN998c6xfvz56enpi6dKlERGxePHimD59ejQ3N0dExIMPPhh/9Vd/FU8++WTceeedsXXr1njttdfimWeeubifBAAAAIDiY8+CBQvi+PHjsWbNmujo6IjZs2dHS0vLwI8wt7e3D7r16JZbboktW7bE6tWr4+GHH45PfOITsX379pg5c+YFX7O8vDyampqG/GoX8KdzxmBkOWMwspwxGFnOGIyskThjJYWL+WwvAAAAAC6pi/frPwAAAABccmIPAAAAQCJiDwAAAEAiYg8AAABAIpdN7Nm4cWPMmDEjKioqora2Nnbv3n3O+T/60Y/iuuuui4qKirj++utj586do7RTGJuKOWObN2+O2267LSZOnBgTJ06M+vr6855J+KAr9u+xP9i6dWuUlJTE/PnzR3aDMMYVe8ZOnjwZy5Yti6lTp0Z5eXl88pOf9O+LcA7FnrH169fHpz71qbjiiiuipqYmli9fHr/97W9Habcwdvz85z+PefPmxbRp06KkpCS2b99+3jW7du2Kz372s1FeXh4f//jH47nnniv6updF7Nm2bVs0NjZGU1NT7N27N2bNmhUNDQ1x7NixIee/+uqrsXDhwrj77rtj3759MX/+/Jg/f3688cYbo7xzGBuKPWO7du2KhQsXxs9+9rNoa2uLmpqauOOOO+Ltt98e5Z3D2FDsGfuDI0eOxNe//vW47bbbRmmnMDYVe8Z6e3vjC1/4Qhw5ciSef/75OHToUGzevDmmT58+yjuHsaHYM7Zly5ZYuXJlNDU1xYEDB+LZZ5+Nbdu2xcMPPzzKO4fLX09PT8yaNSs2btx4QfN/+ctfxp133hm333577N+/Px566KG455574qWXXirqupfFo9dra2vjpptuig0bNkRERH9/f9TU1MQDDzwQK1euPGP+ggULoqenJ1588cWBsc997nMxe/bs2LRp06jtG8aKYs/YH+vr64uJEyfGhg0bYvHixSO9XRhzhnPG+vr64i//8i/j7/7u7+I///M/4+TJkxf0X3rgg6jYM7Zp06b47ne/GwcPHozx48eP9nZhzCn2jN1///1x4MCBaG1tHRj7h3/4h/jv//7veOWVV0Zt3zDWlJSUxAsvvHDOO7pXrFgRO3bsGHQzy5e//OU4efJktLS0XPC1LvmdPb29vbFnz56or68fGCstLY36+vpoa2sbck1bW9ug+RERDQ0NZ50PH2TDOWN/7N1334333nsvrrnmmpHaJoxZwz1j3/zmN2Py5Mlx9913j8Y2Ycwazhn7yU9+EnV1dbFs2bKorq6OmTNnxtq1a6Ovr2+0tg1jxnDO2C233BJ79uwZ+KrX4cOHY+fOnfHFL35xVPYMmV2s3jHuYm5qOE6cOBF9fX1RXV09aLy6ujoOHjw45JqOjo4h53d0dIzYPmGsGs4Z+2MrVqyIadOmnfEPHWB4Z+yVV16JZ599Nvbv3z8KO4SxbThn7PDhw/Ef//Ef8ZWvfCV27twZb731Vvz93/99vPfee9HU1DQa24YxYzhn7K677ooTJ07E5z//+SgUCvG73/0u7rvvPl/jgovgbL2ju7s7fvOb38QVV1xxQe9zye/sAS5v69ati61bt8YLL7wQFRUVl3o7MOadOnUqFi1aFJs3b45JkyZd6u1ASv39/TF58uR45plnYs6cObFgwYJ45JFHfN0fLpJdu3bF2rVr4+mnn469e/fGj3/849ixY0c8/vjjl3prwP/nkt/ZM2nSpCgrK4vOzs5B452dnTFlypQh10yZMqWo+fBBNpwz9gdPPPFErFu3Ln7605/GDTfcMJLbhDGr2DP2i1/8Io4cORLz5s0bGOvv74+IiHHjxsWhQ4fi2muvHdlNwxgynL/Hpk6dGuPHj4+ysrKBsU9/+tPR0dERvb29MWHChBHdM4wlwzljjz76aCxatCjuueeeiIi4/vrro6enJ+6999545JFHorTUPQUwXGfrHZWVlRd8V0/EZXBnz4QJE2LOnDmDftyrv78/Wltbo66ubsg1dXV1g+ZHRLz88stnnQ8fZMM5YxER3/nOd+Lxxx+PlpaWmDt37mhsFcakYs/YddddF6+//nrs379/4PXXf/3XA09cqKmpGc3tw2VvOH+P3XrrrfHWW28NhNSIiDfffDOmTp0q9MAfGc4Ze/fdd88IOn+Iq5fB839gTLtovaNwGdi6dWuhvLy88NxzzxX+53/+p3DvvfcWrr766kJHR0ehUCgUFi1aVFi5cuXA/P/6r/8qjBs3rvDEE08UDhw4UGhqaiqMHz++8Prrr1+qjwCXtWLP2Lp16woTJkwoPP/884V33nln4HXq1KlL9RHgslbsGftjS5YsKfzN3/zNKO0Wxp5iz1h7e3vhqquuKtx///2FQ4cOFV588cXC5MmTC9/61rcu1UeAy1qxZ6ypqalw1VVXFf71X/+1cPjw4cK///u/F6699trC3/7t316qjwCXrVOnThX27dtX2LdvXyEiCk899VRh3759hV/96leFQqFQWLlyZWHRokUD8w8fPly48sorC//4j/9YOHDgQGHjxo2FsrKyQktLS1HXveRf44r4/aPUjx8/HmvWrImOjo6YPXt2tLS0DPwoUXt7+6ByfMstt8SWLVti9erV8fDDD8cnPvGJ2L59e8ycOfNSfQS4rBV7xr7//e9Hb29vfOlLXxr0Pk1NTfGNb3xjNLcOY0KxZwwoTrFnrKamJl566aVYvnx53HDDDTF9+vR48MEHY8WKFZfqI8Blrdgztnr16igpKYnVq1fH22+/HR/+8Idj3rx58e1vf/tSfQS4bL322mtx++23D/y5sbExIiKWLFkSzz33XLzzzjvR3t4+8L9/7GMfix07dsTy5cvje9/7XnzkIx+JH/zgB9HQ0FDUdUsKBffZAQAAAGThPzMCAAAAJCL2AAAAACQi9gAAAAAkIvYAAAAAJCL2AAAAACQi9gAAAAAkIvYAAAAAJCL2AAAAACQi9gAAAAAkIvYAAAAAJCL2AAAAACQi9gAAAAAk8v8AK+eLW+jARaMAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 1400x700 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "### plot the results\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "# plot a line plot with error bars for pearson correlation\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(14, 7))\n",
    "sns.lineplot(x='feature_size', y='corr', data=all_df, ax=ax)\n",
    "ax.set_title(f'Pearson correlation vs feature size')\n",
    "ax.set_xlabel('feature size')\n",
    "ax.set_ylabel('Pearson correlation')\n",
    "# enlarge the tick labels\n",
    "ax.tick_params(axis='x', labelsize=14)\n",
    "ax.tick_params(axis='y', labelsize=14)\n",
    "# enlarge the xlabel and ylabel\n",
    "ax.xaxis.label.set_size(16)\n",
    "ax.yaxis.label.set_size(16)\n",
    "# enlarge the title\n",
    "ax.title.set_fontsize(20)\n",
    "plt.savefig(save_results_path+'feature_size_pearson_corr.png', dpi=300, bbox_inches='tight')\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
